### PR TITLE
Check for possible errors when sending mail

### PR DESCRIPTION
**Pull request by mbabker**
#### Summary of Changes

As of 3.6 JMail will return boolean values on error states from certain methods.  We will try to detect those and implement appropriate error detection/handling with this pull request.

Changes are made in the following places:
- The admin users component's mail model
- The frontend contact component's contact controller - when a contact form is submitted
- The install application's email task controller
- The update notification plugin
- The Joomla User plugin - onAfterUserSave event to send a message when a new user is created
#### Testing Instructions

In these locations, find a way to trigger an error with sending a message (the checks are specifically against methods adding email addresses as PHPMailer validates these, so using invalid addresses may do the trick assuming nothing validates the addresses beforehand).  Instead of attempting to send the message, depending on what the method was already doing for error handling either sending will be skipped completely or the send will be cancelled with an error message raised.

### CODE REVIEW

**Comment by mbabker** [comment]
Less possible fatal errors or errors in general regarding mail sending is a good thing for everyone!

File: administrator/components/com_users/models/mail.php [hunk 1]
```diff
@@ -160,20 +160,46 @@ public function send()
...
-			$mailer->addRecipient($app->get('mailfrom'));
+			if (!$mailer->addBcc($rows))
+			{
+				$app->setUserState('com_users.display.mail.data', $data);
+				$this->setError(JText::_('COM_USERS_MAIL_THE_MAIL_COULD_NOT_BE_SENT'));
+
+				return false;
+			}
+
+			if ($mailer->addRecipient($app->get('mailfrom')))
```
Comment (zero-24) [inline administrator/components/com_users/models/mail.php]: @mbabker can you double check this line? I would expect the following else it fails if it shoud not?

`if (!$mailer->addRecipient($app->get('mailfrom')))`

**Comment by andrepereiradasilva** [comment]
@mbabker i tried to send a mail from contact form and no matter what i do it always say: `Thank you for your email.`

I changes the vars directly in the code to generate an error, but no error:
```php
$contact->email_to = 'aaa';
$email = 'aaa';
$mailfrom = 'aaa';
```
Here https://github.com/joomla/joomla-cms/pull/11230/files#diff-7178e53b18e4d0dc8db925e8b1be7906R202

**Comment by mbabker** [comment]
The check to decide whether the sent message should be shown was checking if an `Exception`

**Comment by mbabker** [comment]
The check to decide whether the sent message should be shown was checking if an `Exception` was returned, which the send method isn't returning (it's using boolean values).  Updated the check.

**Comment by andrepereiradasilva** [comment]
ok, will restest when i have time. thanks

**Comment by mbabker** [comment]
This isn't going anywhere, but I am.

**Comment by rvbgnu** [comment]
The more I read through PR and comments, the more I see that Email issues are going to stay for while. I don't get it, really :-\

**Comment by mbabker** [comment]
All this PR did was improve the error detection from the PHPMailer API when configuring the message to be sent.  Issues related to the actual sending process can't be fixed in Joomla core without subclassing/forking PHPMailer.

**Comment by rvbgnu** [comment]
Thanks Michael. It is become clearer to me know. The underlaying errors, either ignored, or not reported in the upper class, and the lack of error handling from JMail is difficult to work out.

I'm trying to figure out some random errors on a live website. It is very frustrating not being able to find why the default Joomla contact form is not working, only sometimes. I've read also this closed PR https://github.com/joomla/joomla-cms/pull/13339#issuecomment-277802337

What should I add to the code, to gather error logs?

If I can help adding and testing the code, I would do, and having a large scale live context to validate it ;-)

**Comment by mbabker** [comment]
Turn on debug mode, turn on the debug plugin, make sure it is logging at least the mail category.  Since 3.6.0 `JMail` has been configured in a way to turn on PHPMailer's SMTPDebug option and have a callback to log error messages.

**Comment by rvbgnu** [comment]
I've read about JMail and SMTPDebug, that's great. But I've learned again about debug plugin improvements, thanks a million Michael!

I've found the categories allowed, and their syntax. I can leave all, but it will be very verbose on a LIVE website =-0
What should I use for SMTP, PHPMailer and Joomla contact form?

**Pull request closed without merging**